### PR TITLE
feat: support local pending in RPC

### PR DESCRIPTION
**Pull request by Mirko-von-Leipzig**
This is quite a large PR but I have tried my best to make the commits sensible. I would recommend reviewing by commit at least until the last few which are the events related changes.

This PR does not include `starknet_call` and `starknet_feeEstimate` as those are still brewing in #462.

The most complex code is the `get_events` related code, which could use some sharp eyeballs.

### CODE REVIEW

File: crates/pathfinder/src/rpc/api.rs [hunk 1]
```diff
@@ -1011,68 +1011,235 @@ impl RpcApi {
...
+        //  b) if full page           -> return page
+        //  c) else if partially full -> append events from start of pending
+        //  d) else (page is empty):
+        //      i) query database for event count
+        //     ii) query pending data using count for paging into pending events
+
+        use BlockId::*;
+
         let storage = self.storage.clone();
         let span = tracing::Span::current();
```
Comment (kkovaacs) [inline crates/pathfinder/src/rpc/api.rs:1099]: Why aren't we keeping this and `span.enter()` closer together?

File: crates/pathfinder/src/storage/state.rs [hunk 2]
```diff
@@ -817,12 +835,18 @@ impl StarknetEventsTable {
...
 
-        let query = format!(
-            "{} WHERE {}",
-            base_query,
-            where_statement_parts.join(" AND ")
-        );
-
+        let query = if where_statement_parts.is_empty() {
+            format!(
+                "{} ORDER BY block_number, transaction_idx, starknet_events.idx",
```
Comment (kkovaacs) [inline crates/pathfinder/src/storage/state.rs:840]: Why do we need the `ORDER BY` clause here?

File: crates/pathfinder/src/storage/state.rs [hunk 3]
```diff
@@ -774,13 +774,17 @@ impl StarknetEventsTable {
...
         contract_address: Option<ContractAddress>,
         keys: Vec<EventKey>,
     ) -> anyhow::Result<usize> {
         let mut base_query =
-        r#"SELECT COUNT(1)
+        r#"SELECT COUNT(1),
+           block_number,
+           starknet_transactions.idx as transaction_idx,
+           from_address,
+           starknet_events.keys as keys
```
Comment (kkovaacs) [inline crates/pathfinder/src/storage/state.rs:787]: Why are these required? We only seem to use the first column from the result.

Comment (Mirko-von-Leipzig) [reply crates/pathfinder/src/storage/state.rs]: I think they were required by the joins. Which are themselves not required.

File: crates/pathfinder/src/rpc.rs [hunk 4]
```diff
@@ -867,9 +867,20 @@ mod tests {
...
+            let pending_data = create_pending_data(storage.clone()).await;
+            let pending_block = pending_data.block().await.unwrap();
+            let sequencer = Client::new(Chain::Goerli).unwrap();
+            let sync_state = Arc::new(SyncState::default());
+            let api = RpcApi::new(storage, sequencer, Chain::Goerli, sync_state)
+                .with_pending_data(pending_data);
+            let (__handle, addr) = run_server(*LOCALHOST, api).await.unwrap();
             let params = rpc_params!(BlockId::Pending);
-
-            check_result(params, move |_| {}).await;
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply crates/pathfinder/src/rpc.rs:706]: I added the pending block as an input to the test function.

**Review by kkovaacs — APPROVED** [review]

File: crates/pathfinder/src/rpc/api.rs [hunk 5]
```diff
@@ -745,7 +746,33 @@ impl RpcApi {
...
+                                    .connection()
+                                    .context("Opening database connection")
+                                    .map_err(internal_server_error)?;
+                                let tx = db
+                                    .transaction()
+                                    .context("Creating database transaction")
+                                    .map_err(internal_server_error)?;
+                                let code = ContractCodeTable::get_class(&tx, class_hash)
+                                    .context("Fetching code from database");
+                                dbg!(&code);
```
Comment (CHr15F0x) [inline crates/pathfinder/src/rpc/api.rs:758]: I guess this was for development?

**Review by CHr15F0x — APPROVED** [review]
LGTM, this was a lot of work :+1: 
----
noticed two `FIXME: add support for pending` that should be removed, which gh won't let me comment on directly.
